### PR TITLE
Raise ApplicationEvent.PreSendRequestHeaders when writing starts

### LINKED ISSUE #514: PreSendRequestHeaders and PreSendRequestContent should be fired when content is being sent
If buffering is disabled and something is written. The framework `HttpApplication` will short circuit the request pipeline and jump to `RequestNotification.SendReponse` and raise `PreSendRequestHeaders` and `PreSendRequestContent` events.
In comparison systemweb-adapter will skip all the next stages and send the response.

`HttpReponse.Flush()` will expliexplicitly set `RequestNotification = SendReponse`. The `StepManager` will execute it. Following is the callstack.
```
>	System.Web.dll!System.Web.HttpContext.CurrentNotification.set(System.Web.RequestNotification **value = SendResponse**) Line 993	C#
 	System.Web.dll!System.Web.HttpRuntime.ProcessRequestNotificationPrivate(System.Web.Hosting.IIS7WorkerRequest wr, System.Web.HttpContext context) Line 1374	C#
 	System.Web.dll!System.Web.Hosting.PipelineRuntime.ProcessRequestNotificationHelper(System.IntPtr rootedObjectsPointer, System.IntPtr nativeRequestContext, System.IntPtr moduleData, int flags) Line 461	C#
 	System.Web.dll!System.Web.Hosting.PipelineRuntime.ProcessRequestNotification(System.IntPtr rootedObjectsPointer, System.IntPtr nativeRequestContext, System.IntPtr moduleData, int flags) Line 386	C#
 	[Native to Managed Transition]	
 	[Managed to Native Transition]	
 	System.Web.dll!System.Web.Hosting.IIS7WorkerRequest.ExplicitFlush() Line 1729	C#
 	System.Web.dll!System.Web.HttpResponse.Flush(bool finalFlush, bool async) Line 1238	C#
 	System.Web.dll!System.Web.HttpWriter.WriteFromStream(byte[] data, int offset, int size) Line 341	C#

```

_Originally posted by @TaoziZ03 in https://github.com/dotnet/systemweb-adapters/issues/509#issuecomment-2224873982_

### PR DESCRIPTION
**Pull request by twsouthwick**
Fixes #514

### CODE REVIEW

**Comment by twsouthwick** [comment]
This is stacked against #509 since it requires the ability to disable buffering to test it.

**Review by TaoziZ03 — APPROVED** [review]
Looks good, thanks!

File: test/Microsoft.AspNetCore.SystemWebAdapters.CoreServices.Tests/Microsoft.AspNetCore.SystemWebAdapters.CoreServices.Tests.csproj [hunk 1]
```diff
@@ -7,6 +7,7 @@
     <PackageReference Include="AutoFixture" Version="4.15.0" />
     <PackageReference Include="Moq" Version="4.16.1" />
     <PackageReference Include="Microsoft.AspNetCore.TestHost" Version="6.0.13" />
+    <PackageReference Include="System.Collections.Immutable" Version="8.0.0" />
```
Comment (joperezr) [inline test/Microsoft.AspNetCore.SystemWebAdapters.CoreServices.Tests/Microsoft.AspNetCore.SystemWebAdapters.CoreServices.Tests.csproj:10]: Unrelated to your PR, but we should think about setting up dependabot or some other way to automatically bump these dependencies in the future.

**Review by joperezr — APPROVED** [review]
Looks good, thanks Taylor!

**Pull request merged**